packages/medusa/src/types/product.ts

```diff
@@ -123,3 +127,119 @@ export class FilterableProductTypeProps {
   @IsOptional()
   q?: string
 }
+
+/**
+ * Service Level DTOs
+ */
+
+export type CreateProductInput = {
+  title: string
+  subtitle?: string
+  profile_id?: string
+  description?: string
+  is_giftcard?: boolean
+  discountable?: boolean
+  images?: string[]
+  thumbnail?: string
+  handle?: string
+  status?: ProductStatus
+  type?: CreateProductProductTypeInput
+  collection_id?: string
+  tags?: CreateProductProductTagInput[]
+  options?: CreateProductProductOption[]
+  variants?: CreateProductProductVariantInput[]
+  weight?: number
+  length?: number
+  height?: number
+  width?: number
+  hs_code?: string
+  origin_country?: string
+  mid_code?: string
+  material?: string
+  metadata?: Record<string, unknown>
+}
+
+export type CreateProductProductTagInput = {
+  id?: string
+  value: string
+}
+
+export type CreateProductProductTypeInput = {
+  id?: string
+  value: string
+}
+
+export type CreateProductProductVariantInput = {
+  title: string
+  sku?: string
+  ean?: string
+  upc?: string
+  barcode?: string
+  hs_code?: string
+  inventory_quantity?: number
+  allow_backorder?: boolean
+  manage_inventory?: boolean
+  weight?: number
+  length?: number
+  height?: number
+  width?: number
+  origin_country?: string
+  mid_code?: string
+  material?: string
+  metadata?: object
+  prices?: CreateProductProductVariantPriceInput[]
+  options?: { value: string }[]
+}
+
+export type UpdateProductProductVariantDTO = {
+  id?: string
+  title?: string
+  sku?: string
+  ean?: string
+  upc?: string
+  barcode?: string
+  hs_code?: string
+  inventory_quantity?: number
+  allow_backorder?: boolean
+  manage_inventory?: boolean
+  weight?: number
+  length?: number
+  height?: number
+  width?: number
+  origin_country?: string
+  mid_code?: string
+  material?: string
+  metadata?: object
+  prices?: CreateProductProductVariantPriceInput[]
+  options?: { value: string; option_id: string }[]
+}
+
+export type CreateProductProductOption = {
+  title: string
+}
+
+export type CreateProductProductVariantPriceInput = {
+  region_id?: string
+  currency_code?: string
+  amount: number
+  min_quantity?: number
+  max_quantity?: number
+}
+
+export type UpdateProductInput = Omit<
+  Partial<CreateProductInput>,
+  "variants"
+> & {
+  variants?: UpdateProductProductVariantDTO[]
+}
+
+export type ProductOptionInput = {
+  title: string
+  values?: ProductOptionValue[]
+}
+
+export type FindProductConfig = FindConfig<Product> & PriceListLoadConfig
+
+export type ProductSelector = Omit<Selector<Product>, "tags"> & {
```

Comment:
You have a point. The reason why I came up with this in the first place is because of how the `queryProducts` method is implemented in the `ProductRepository`. 
https://github.com/medusajs/medusa/blob/f2580e8e834b39c0f2aee06a0d066d0830f412de/packages/medusa/src/repositories/product.ts#L70-L74

By line 71, this is what the `tags` type looks like:
```
ProductTag[] | DateComparisonOperator | StringComparisonOperator | NumericalComparisonOperator | ... 4 more ... | undefined
```
however, the code assumes that the `tags` variable will be of type `FindOperator`.

I thought of 2 ways to tackle this one:
1. Either you add a check in the `if` statement in line 70 to make sure that `tags instanceof FindProduct`
2. Tackle the problem from the other side by trying to remove the unnecessary types when passing them down. Also, the repo method argument can be changed to a more "narrow" type that would reflect the implementation more, hence why the new `ProductSelector` type which gives a narrower representation of the implementation. 

Downside with option 2 imo is that it pushed me down a rabbit hole because the change breaks some types elsewhere and as discussed previously, I think this PR is getting out of hands already haha 😅 . So I went with option 1 in the latest commit.

Reply:
😅 what i meant was to only remove it from the service, in the repository you can let it as it is and remove the check on tags FindOperator.